server/proxy/grpcproxy/cluster.go

```diff
@@ -23,7 +23,7 @@ import (
 
 	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
 	"go.etcd.io/etcd/api/v3/v3rpc/rpctypes"
-	"go.etcd.io/etcd/client/v3"
+	clientv3 "go.etcd.io/etcd/client/v3"
```

Comment:
Reasons:

- https://github.com/etcd-io/etcd/blob/4accc34c98e484b5a142cf367a76bfd48d42c426/server/proxy/grpcproxy/cluster.go#L39
- https://github.com/etcd-io/etcd/blob/4accc34c98e484b5a142cf367a76bfd48d42c426/server/proxy/grpcproxy/cluster.go#L54

Does this package build as-is? Does Go / `go mod` have some kind of magic / implicit alias-ing rule here?

Reply:
Given the alias only happens 5 times

```
$ git grep -n 'clientv3 "go\.etcd' -- '*.go'
client/v3/naming/endpoints/endpoints.go:6:      clientv3 "go.etcd.io/etcd/client/v3"
client/v3/naming/endpoints/endpoints_impl.go:9: clientv3 "go.etcd.io/etcd/client/v3"
client/v3/naming/resolver/resolver.go:4:        clientv3 "go.etcd.io/etcd/client/v3"
tests/integration/proxy/grpcproxy/register_test.go:21:  clientv3 "go.etcd.io/etcd/client/v3"
tools/etcd-dump-metrics/etcd.go:26:     clientv3 "go.etcd.io/etcd/client/v3"
```

but the alias is (probably) used in 131 files

```
$ git grep -l 'clientv3\.' -- '*.go' | wc -l
     131
```

it seems maybe there is a Go feature I'm not appreciating.

Reply:
Empirically, I just tried with

```go
package main

import (
	"fmt"

	"github.com/jackc/pgx/v4"
)

func main() {
	to := pgx.TxOptions{}
	fmt.Printf("%#v\n", to)
	to = pgxv4.TxOptions{}
	fmt.Printf("%#v\n", to)
}
```

and got a

```
$ go run ./main.go 
# command-line-arguments
./main.go:12:7: undefined: pgxv4
```

so AFAIK, no magic alias feature.

Reply:
```go
package main

import (
	"fmt"

	"go.etcd.io/etcd/client/v3"
)

func main() {
	c := client.AuthRoleAddResponse{}
	fmt.Printf("%#v\n", c)
}
```

produces

```
$ go run ./main.go 
# command-line-arguments
./main.go:6:2: imported and not used: "go.etcd.io/etcd/client/v3" as clientv3
./main.go:10:7: undefined: client
```

and now I finally understand!!

https://github.com/etcd-io/etcd/blob/4accc34c98e484b5a142cf367a76bfd48d42c426/client/v3/doc.go#L106